src/material/datepicker/multi-year-view.ts

```diff
@@ -39,6 +39,43 @@ export const yearsPerPage = 24;
 
 export const yearsPerRow = 4;
 
+/** Gets remainder that is non-negative, even if first number is negative */
+function euclideanModulo (a: number, b: number): number {
+  return (a % b + b) % b;
+}
+
+/**
+ * We pick a "starting" year such that either the maximum year would be at the end
+ * or the minimum year would be at the beginning of a page.
+ */
+function getStartingYear(minDate: any, maxDate: any): number {
+  let startingYear = 0;
+  if (maxDate) {
+    const maxYear = maxDate.getFullYear();
```

Comment:
It looks like these helper functions aren't going through the adapter any more, so they're assuming native JS `Date` (here and in `getActiveOffset`). You could make these `private` class methods so you still have access to the adapter. 

You also would want to avoid using `any` in the types because it masks issues like this

Reply:
@jelbourn  Thanks, sorry for MIA the past week. 

1. So I moved getStartingYear, getActiveOffset, and isSameMultiYearView into the MatMultiYearView class to access the date adapter. Which of the following approach do you prefer?

a. To avoid repetition, share getActiveOffset and isSameMultiYearView with MatCalendar: But these two methods cannot be private (and public_api_guard/material/datepicker.d.ts updated?), so that MatCalendar can access them via ViewChildren for the Header's periodButtonText and isSameView check.

b.  Repeat getStartingYear, getActiveOffset, and isSameMultiYearView as private methods for the MatCalendarHeader as well (like how private _getValidDateOrNull is repeated across MatCalendar and MatMultiYearView).

2.  _getActiveCell in MatMultiYearView is essentially (and is calling) getActiveOffset, but the two names sound relevant for their respective purposes, so I left them as they are.